MiniBrowser address bar immediately relinquishes focus when tab focusing out from web content

beidson opened the pull request:
#### 24de837345794a299744beac26de7e2000253de8
<pre>
MiniBrowser address bar immediately relinquishes focus when tab focusing out from web content
<a href="https://rdar.apple.com/160792178">rdar://160792178</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=299392">https://bugs.webkit.org/show_bug.cgi?id=299392</a>

Reviewed by Abrar Rahman Protyasha.

If we relinquish focus to the chrome client, we should also mark the event as handled.
Otherwise, WebKit will automatically forward it along to AppKit for further processing.

Test: Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm

* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::findAndFocusElementInDocumentOrderStartingWithFrame):
* Source/WebCore/page/FocusController.h:

* Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
(-[TestNSTextView keyDown:]):
(-[TestNSTextView acceptsFirstResponder]):
(-[TestNSTextView canBecomeKeyView]):
(-[TestNSTextView becomeFirstResponder]):
(-[TestNSTextView resignFirstResponder]):
(TestWebKitAPI::AdvanceFocusRelinquishToChrome)):
(TestWebKitAPI::TEST(FocusWebView, DoNotFocusWebViewWhenUnparented)): Deleted.

Canonical link: <a href="https://commits.webkit.org/300442@main">https://commits.webkit.org/300442@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64c3b716654f37d98e2e289e610ac5949346c1dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122543 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42251 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32936 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129152 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74644 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/490c6348-0f37-4955-a52f-436d6fd4fdc1) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124419 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50844 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93149 "5 flakes 16 failures") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125495 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34272 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109722 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73796 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlot (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/16822ce8-3167-49ac-835d-ffde852b46cd) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33255 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27879 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72636 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103929 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28089 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131878 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49484 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37663 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101682 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49859 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105942 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101550 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25782 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46927 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25077 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46254 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49342 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55091 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48810 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52161 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50491 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
```diff
@@ -37,8 +38,128 @@
+
+- (void)keyDown:(NSEvent *)event
+{
+    NSLog(@"keyDown!!!!");
```
Stray log!

beidson replied:
Removed locally but not pushed - yikes!

aprotyas commented on Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
```diff
@@ -37,8 +38,127 @@
+    RetainPtr<TestUIDelegate> uiDelegate;
+};
+
+static WebViewAndDelegates makeWebViewAndDelegates(HTTPServer& server, bool enable = true)
```
Seems like `enable` isn't being used for anything?

beidson replied:
Yah, carried over from code it was cribbed from. Will remove.

aprotyas commented on Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
```diff
@@ -37,8 +38,127 @@
+static WebViewAndDelegates makeWebViewAndDelegates(HTTPServer& server, bool enable = true)
+{
+    RetainPtr configuration = server.httpsProxyConfiguration();
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 400, 400) configuration:configuration.get()]);
```
```suggestion
    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 400, 400) configuration:configuration.get()]);
```

aprotyas commented on Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
```diff
@@ -37,8 +38,127 @@
+    RetainPtr configuration = server.httpsProxyConfiguration();
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 400, 400) configuration:configuration.get()]);
+
+    auto navigationDelegate = adoptNS([TestNavigationDelegate new]);
```
```suggestion
    RetainPtr navigationDelegate = adoptNS([TestNavigationDelegate new]);
```

aprotyas commented on Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
```diff
@@ -37,8 +38,127 @@
+    auto webViewAndDelegates = makeWebViewAndDelegates(server);
+    auto webView = WTFMove(webViewAndDelegates.webView);
+    auto navigationDelegate = WTFMove(webViewAndDelegates.navigationDelegate);
+    auto uiDelegate = WTFMove(webViewAndDelegates.uiDelegate);
```
`RetainPtr` for the types?

aprotyas commented on Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
```diff
@@ -37,8 +38,127 @@
+    [[[webView.get() window] contentView] addSubview:textField.get()];
+    [[webView.get() window] setFrame:newWindowFrame display:YES];
+
+    [webView.get().window makeKeyWindow];
```
You can skip the `.get()` when you use the selector syntax
```suggestion
    [[[webView window] contentView] addSubview:textField.get()];
    [[webView window] setFrame:newWindowFrame display:YES];

    [[webView window] makeKeyWindow];
```

aprotyas commented on Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
```diff
@@ -37,8 +38,127 @@
+};
+
+
+// FIXME: To enable, need `typeCharacter:` support for TestWKWebView on iOS
```
Btw, if you intended to run this test on iOS, there are ways to do something typeCharacter: adjacent, here's a `pressSpacebar` example from UnifiedPDFTests.mm

```
    auto pressSpacebar = ^(void(^completionHandler)(void)) {
        RetainPtr firstWebEvent = adoptNS([[WebEvent alloc] initWithKeyEventType:WebEventKeyDown timeStamp:CFAbsoluteTimeGetCurrent() characters:@" " charactersIgnoringModifiers:@" " modifiers:0 isRepeating:NO withFlags:0 withInputManagerHint:nil keyCode:0 isTabKey:NO]);

        RetainPtr secondWebEvent = adoptNS([[WebEvent alloc] initWithKeyEventType:WebEventKeyUp timeStamp:CFAbsoluteTimeGetCurrent() characters:@" " charactersIgnoringModifiers:@" " modifiers:0 isRepeating:NO withFlags:0 withInputManagerHint:nil keyCode:0 isTabKey:NO]);

        [webView handleKeyEvent:firstWebEvent.get() completion:^(WebEvent *theEvent, BOOL wasHandled) {
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), mainDispatchQueueSingleton(), ^{
                [webView handleKeyEvent:secondWebEvent.get() completion:^(WebEvent *theEvent, BOOL wasHandled) {
                    completionHandler();
                }];
            });
        }];
    };
```

beidson replied:
For context: This pattern was almost directly cribbed from other site isolation tests, where the "type a character on iOS" problem originates from.

We should solve it, but all of them at once, and not now in this one 😁

beidson committed:
MiniBrowser address bar immediately relinquishes focus when tab focusing out from web content
rdar://160792178
https://bugs.webkit.org/show_bug.cgi?id=299392

Reviewed by Abrar Rahman Protyasha.

If we relinquish focus to the chrome client, we should also mark the event as handled.
Otherwise, WebKit will automatically forward it along to AppKit for further processing.

Test: Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm

* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::findAndFocusElementInDocumentOrderStartingWithFrame):
* Source/WebCore/page/FocusController.h:

* Tools/TestWebKitAPI/Tests/mac/FocusWebView.mm:
(-[TestNSTextView keyDown:]):
(-[TestNSTextView acceptsFirstResponder]):
(-[TestNSTextView canBecomeKeyView]):
(-[TestNSTextView becomeFirstResponder]):
(-[TestNSTextView resignFirstResponder]):
(TestWebKitAPI::AdvanceFocusRelinquishToChrome)):
(TestWebKitAPI::TEST(FocusWebView, DoNotFocusWebViewWhenUnparented)): Deleted.

Canonical link: https://commits.webkit.org/300442@main
